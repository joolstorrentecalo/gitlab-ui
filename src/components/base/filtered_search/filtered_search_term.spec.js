import { nextTick } from 'vue';
import { shallowMount } from '@vue/test-utils';
import GlFilteredSearchSuggestion from './filtered_search_suggestion.vue';
import FilteredSearchTerm from './filtered_search_term.vue';

const availableTokens = [
  { type: 'foo', title: 'test1-foo', token: 'stub', icon: 'eye' },
  { type: 'bar', title: 'test2-bar', token: 'stub', icon: 'eye' },
  { type: 'baz', title: 'test1-baz', token: 'stub', icon: 'eye' },
];

describe('Filtered search term', () => {
  let wrapper;

  const searchInputAttributes = { 'data-qa-selector': 'foo-bar' };

  const defaultProps = {
    availableTokens: [],
  };

  const segmentStub = {
    name: 'gl-filtered-search-token-segment-stub',
    template: '<div><slot name="view"></slot><slot name="suggestions"></slot></div>',
    props: ['searchInputAttributes', 'isLastToken'],
  };

  const createComponent = (props) => {
    wrapper = shallowMount(FilteredSearchTerm, {
      propsData: { ...defaultProps, ...props },
      stubs: {
        'gl-filtered-search-token-segment': segmentStub,
      },
    });
  };

  const findTokenSegmentComponent = () => wrapper.findComponent(segmentStub);

  it('renders value in inactive mode', () => {
    createComponent({ value: { data: 'test-value' } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders input with value in active mode', () => {
    createComponent({ value: { data: 'test-value' }, active: true });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders input with placeholder if placeholder prop is provided', () => {
    createComponent({ placeholder: 'placeholder-stub' });
    expect(wrapper.find('input').attributes('placeholder')).toBe('placeholder-stub');
  });

  it('filters suggestions by input', async () => {
    createComponent({ availableTokens, active: true, value: { data: 'test1' } });

    await nextTick();

    expect(wrapper.findAllComponents(GlFilteredSearchSuggestion)).toHaveLength(2);
  });

  it.each`
    originalEvent   | emittedEvent
    ${'activate'}   | ${'activate'}
    ${'deactivate'} | ${'deactivate'}
    ${'split'}      | ${'split'}
    ${'submit'}     | ${'submit'}
    ${'complete'}   | ${'replace'}
    ${'backspace'}  | ${'destroy'}
  `(
    'emits $emittedEvent when token segment emits $originalEvent',
    async ({ originalEvent, emittedEvent }) => {
      createComponent({ active: true, value: { data: 'something' } });

      findTokenSegmentComponent().vm.$emit(originalEvent);

      await nextTick();

      expect(wrapper.emitted()[emittedEvent]).toHaveLength(1);
    }
  );

  it('passes `searchInputAttributes` and `isLastToken` prop to `GlFilteredSearchTokenSegment`', () => {
    const isLastToken = true;

    createComponent({
      value: { data: 'something' },
      searchInputAttributes,
      isLastToken,
    });

    expect(findTokenSegmentComponent().props()).toEqual({
      searchInputAttributes,
      isLastToken,
    });
  });

  it('adds `searchInputAttributes` prop to search term input', () => {
    createComponent({
      placeholder: 'placeholder-stub',
      searchInputAttributes,
    });

    expect(wrapper.find('input').attributes('data-qa-selector')).toBe(
      searchInputAttributes['data-qa-selector']
    );
  });
});
