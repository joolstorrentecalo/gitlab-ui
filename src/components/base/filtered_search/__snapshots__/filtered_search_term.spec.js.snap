// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filtered search term renders input with value in active mode 1`] = `
<div
  class="gl-h-auto gl-filtered-search-term"
  data-testid="filtered-search-term"
>
  <div
    active="true"
    class="gl-filtered-search-term-token"
    cursor-position="end"
    value="test-value"
  >
    test-value
  </div>
</div>
`;

exports[`Filtered search term renders value in inactive mode 1`] = `
<div
  class="gl-h-auto gl-filtered-search-term"
  data-testid="filtered-search-term"
>
  <div
    class="gl-filtered-search-term-token"
    cursor-position="end"
    value="test-value"
  >
    test-value
  </div>
</div>
`;
